src/main/java/graphql/execution/instrumentation/dataloader/PerLevelDataLoaderDispatchStrategy.java

```diff
@@ -275,9 +310,89 @@ private boolean levelReady(int level) {
     }
 
     void dispatch(int level) {
-        DataLoaderRegistry dataLoaderRegistry = executionContext.getDataLoaderRegistry();
-        dataLoaderRegistry.dispatchAll();
+        if (callStack.levelToDFEWithDataLoaderCF.size() > 0) {
+            dispatchDLCFImpl(callStack.levelToDFEWithDataLoaderCF.get(level));
+        } else {
+            DataLoaderRegistry dataLoaderRegistry = executionContext.getDataLoaderRegistry();
+            dataLoaderRegistry.dispatchAll();
+        }
     }
 
+
+    public void dispatchDLCFImpl(Set<DataFetchingEnvironment> dfeToDispatchSet) {
+
+        // filter out all DataLoaderCFS that are matching the fields we want to dispatch
+        List<DataLoaderCF<?>> relevantDataLoaderCFs = new ArrayList<>();
+        for (DataLoaderCF<?> dataLoaderCF : callStack.allDataLoaderCF) {
+            if (dfeToDispatchSet.contains(dataLoaderCF.dfe)) {
+                relevantDataLoaderCFs.add(dataLoaderCF);
+            }
+        }
+        // we are cleaning up the list of all DataLoadersCFs
+        callStack.allDataLoaderCF.removeAll(relevantDataLoaderCFs);
+
+        // means we are all done dispatching the fields
+        if (relevantDataLoaderCFs.size() == 0) {
+            callStack.fieldsFinishedDispatching.addAll(dfeToDispatchSet);
+            return;
+        }
+        // we are dispatching all data loaders and waiting for all dataLoaderCFs to complete
+        // and to finish their sync actions
+        CountDownLatch countDownLatch = new CountDownLatch(relevantDataLoaderCFs.size());
+        for (DataLoaderCF dlCF : relevantDataLoaderCFs) {
+            dlCF.latch = countDownLatch;
```

Comment:
There is no need for dlCF.latch = countDownLatch and hence mutability

Since we have the dlDF we can do `dlCF.whenComplete( v,e -> {countDownLatch.countDown()}}`

Which is what it does on the inside right now.

Reply:
(or does the countdown latch HAVE to decrement before the value is returned ???)